### PR TITLE
VPN-3331 - WIP: Crash reporting for all! 

### PR DESCRIPTION
**Pull request by strseb**
- Fetch Crash Report info from Guardian
- Enable Crashreporting for Linux

## Description

    Describe your changes

## Reference

    i.e Jira or Github issue URL

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+bool TaskSentryConfig::handleNetworkResponse(const QByteArray& data) {
+  auto doc = QJsonDocument::fromJson(data);
+  if (!doc.isObject()) {
+    logger.error() << "Failed to get Sentry info";
+    return false;
+  }
+  // The only thing right now, we don't want to send
+  // crash-data, so we will only check for that.
```
Comment (brizental) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:60]: I don't understand this comment.

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    return false;
+  };
+
+  if (!obj.contains("endpoint")) {
+    logger.error() << "endpoint missing from Sentry info";
+    return false;
+  };
+  auto endpoint = obj["endpoint"].toString();
+  if (endpoint.isNull()) {
+    logger.error() << "endpoint missing from Sentry info";
```
Comment (brizental) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:78]: nit:
```suggestion
    logger.error() << "Sentry provided endpoint is null";
```

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    return false;
+  };
+  auto settings = SettingsHolder::instance();
+  settings->setSentryEndpoint(endpoint);
+  settings->setSentryDSN(dsn);
```
Comment (brizental) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:83]: How come this is not constant

File: src/shared/sentry/sentryadapter.cpp [hunk 4]
```diff
@@ -4,7 +4,10 @@
 
 #include "sentryadapter.h"
 
-#include <QtQml/private/qqmlengine_p.h>
+#ifndef UNIT_TEST
+// This is not found when building tests ... wierd.
```
[the remark this answers is not in the document]
Comment (strseb) [reply src/shared/sentry/sentryadapter.cpp:8]: https://github.com/mozilla-mobile/mozilla-vpn-client/pull/7176/commits/dfe8c8696fbfa2b322e5aced6b30d7a00c785a5d#diff-75169c8db799911a8b0af81ad1c17c24e41a868d224d756d249b7f8b070cee0c

tried that before, did not work :c `auth_test` did not build. In theory given sentry is part of shared-sources and that is an interface target, it "should" build. Also locally that builds too, so i blame something in github ci. 

>How come it's not necessary in tests?
Because i'm also "#ifndef UNIT_TEST" ing the code that calls this

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp (hunk 1 above)
Comment (strseb) [reply src/shared/tasks/sentryconfig/tasksentryconfig.cpp:60]: Me neither, lol wtf

File: src/shared/sentry/sentryadapter.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (brizental) [reply src/shared/sentry/sentryadapter.cpp:8]: Got it. Hum, maybe because the code is not called ever it's being optimized away. Who know, looks correct to not import it in this case anyways.

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp (hunk 3 above)
Comment (strseb) [reply src/shared/tasks/sentryconfig/tasksentryconfig.cpp:81]: Now currently you need to know our sentry creds to enable crash reporting. 
This is not good for linux, as when we build using PPA's we cannot have secrets :) 
But we also don't want that piece of info in our repo. 
That's why we moved that part to guardian and pull it from there.

Comment (brizental) [reply src/shared/tasks/sentryconfig/tasksentryconfig.cpp:81]: Ah ok, so the whole reason for this is Linux essentially? I get not wanting it on our repo, but it could have been a build var otherwise.

Comment (strseb) [reply src/shared/tasks/sentryconfig/tasksentryconfig.cpp:81]: > Ah ok, so the whole reason for this is Linux essentially? 
Yes. :D

Comment (brizental) [reply src/shared/tasks/sentryconfig/tasksentryconfig.cpp:81]: Ok, would you mind adding a comment explaining this on the code? I guess the definition of TaskSentryConfig is the best place for it.

File: src/shared/cmake/sentry.cmake [hunk 5]
```diff
@@ -81,29 +68,40 @@ if( ${_SUPPORTED} GREATER -1 )
...
                         -DANDROID_ABI=${ANDROID_ABI}
                         -DCMAKE_ANDROID_NDK=${ANDROID_NDK_ROOT}
                         -DCMAKE_TOOLCHAIN_FILE=${ANDROID_NDK_ROOT}/build/cmake/android.toolchain.cmake
                         -DSENTRY_BACKEND=inproc
             )
+    
+    endif()
+    if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
+        target_compile_definitions(shared-sources INTERFACE SENTRY_BUILD_STATIC)
+        # Compile Static for apple and link to libsentry.a
```
Comment (brizental) [inline src/shared/cmake/sentry.cmake:89]: This comment is stale.

File: src/shared/sentry/sentryadapter.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (brizental) [reply src/shared/sentry/sentryadapter.cpp:8]: I guess this comment can go though.

File: src/shared/sentry/sentryadapter.cpp [hunk 6]
```diff
@@ -37,17 +41,25 @@ SentryAdapter::SentryAdapter() { MZ_COUNT_CTOR(SentryAdapter); }
...
   if (!Feature::get(Feature::Feature_sentry)->isSupported()) {
     return;
   }
-  if (QString(Constants::SENTRY_DSN_ENDPOINT).isEmpty() ||
-      QString(Constants::SENTRY_ENVELOPE_INGESTION).isEmpty()) {
+  auto settings = SettingsHolder::instance();
+  if (!settings->hasSentryEndpoint() || !settings->hasSentryDSN()) {
+    // If we have no info on where to put crash info, let's
+    // query that and retry later.
     logger.error() << "Sentry failed to init, no sentry config present";
```
Comment (brizental) [inline src/shared/sentry/sentryadapter.cpp:54]: This is not really an error but an expected state when sentry has not been initialized before. Maybe reword and make it not an error?

**Review by brizental — APPROVED** [review]
Just nits. LGTM.

File: src/shared/sentry/sentryadapter.cpp (hunk 4 above)
Comment (Gela) [inline src/shared/sentry/sentryadapter.cpp:8]:
```suggestion
// This is not found when building tests ... weird.
```

File: src/shared/sentry/sentryadapter.cpp [hunk 7]
```diff
@@ -37,17 +41,25 @@ SentryAdapter::SentryAdapter() { MZ_COUNT_CTOR(SentryAdapter); }
...
+    auto t = new TaskSentryConfig();
```
Comment (Gela) [inline src/shared/sentry/sentryadapter.cpp:55]: In this instance it's not as prevalent but generally speaking I try to avoid one character var names
```suggestion
    auto sentryConfig = new TaskSentryConfig();
```

File: src/shared/tasks/sentry/tasksentry.cpp [hunk 8]
```diff
@@ -76,12 +76,20 @@ void TaskSentry::run() {
 
 void TaskSentry::sendRequest() {
   NetworkRequest* request = new NetworkRequest(this, 200);
+  auto settings = SettingsHolder::instance();
+  auto endpoint = settings->sentryEndpoint();
+  auto dsn = settings->sentryDSN();
+  if (endpoint.isNull() || dsn.isNull()) {
+    // We could have not even been initializing sentry without that?!
```
Comment (Gela) [inline src/shared/tasks/sentry/tasksentry.cpp:83]: Should we rephrase a comment like this to something a little more specific? maybe like:
```suggestion
    // We should not be able to initialize Sentry without an endpoint and DSN
```

**Review by Gela — APPROVED** [review]

File: .github/workflows/functional_tests.yaml [hunk 9]
```diff
@@ -24,7 +24,7 @@ jobs:
       - name: Clone repository
         uses: actions/checkout@v3
         with:
-          submodules: 'true'
+          submodules: 'recursive'
```
Comment (vinocher) [inline .github/workflows/functional_tests.yaml:27]: Should docs\dev-setup\index.md be updated to add --recursive to 'git submodule update --init'?

File: src/shared/tasks/sentry/tasksentry.cpp [hunk 10]
```diff
@@ -76,12 +76,20 @@ void TaskSentry::run() {
 
 void TaskSentry::sendRequest() {
   NetworkRequest* request = new NetworkRequest(this, 200);
```
Comment (vinocher) [inline src/shared/tasks/sentry/tasksentry.cpp:78]: Won't this allocation be leaked if the code returns at line 86? Should this line be moved below the new code block?

File: src/shared/tasks/sentryconfig/tasksentryconfig.h [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+#ifndef TASKSENTRYCONFIG_H
+#define TASKSENTRYCONFIG_H
+
+#include <QByteArray>
+#include <QObject>
+
+#include "task.h"
+
+class TaskSentryConfig final : public Task {
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.h:14]: In general, it would be very useful to add a class header comment that summarizes purpose and usage. But that pattern doesn't seem to be used in this code base.

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+void TaskSentryConfig::run() {
+  NetworkRequest* request = new NetworkRequest(this, 200);
+
+  QUrl url(AppConstants::apiBaseUrl());
+  url.setPath("/api/v1/vpn/crashreporting");
+
+  connect(request, &NetworkRequest::requestFailed, this,
+          [this](QNetworkReply::NetworkError error, const QByteArray&) {
+            Q_UNUSED(error);
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:37]: Nit: Instead of Q_UNUSED, couldn't the unused parameter name 'error' be removed?

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+              SentryAdapter::instance()->init();
+            }
+            emit completed();
+          });
+
+  request->get(url);
+}
+
+bool TaskSentryConfig::handleNetworkResponse(const QByteArray& data) {
+  auto doc = QJsonDocument::fromJson(data);
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:54]: Shouldn't doc be null checked to handle fromJson failure?

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+          [this](QNetworkReply::NetworkError error, const QByteArray&) {
+            Q_UNUSED(error);
+            logger.error() << "Failed to get Sentry info";
+            emit completed();
+          });
+  connect(request, &NetworkRequest::requestCompleted, this,
+          [this](const QByteArray& data) {
+            if (handleNetworkResponse(data)) {
+              // We now should have sentry info - let's init
+              SentryAdapter::instance()->init();
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:45]: Consider moving this into handleNetworkResponse for better cohesion, instead of depending on the side effect that handleNetworkResponse has set the sentry config info into SettingsHolder.

File: src/shared/tasks/sentryconfig/tasksentryconfig.h [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+  Q_DISABLE_COPY_MOVE(TaskSentryConfig)
+
+ public:
+  TaskSentryConfig();
+  ~TaskSentryConfig();
+
+  void run() override;
+
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.h:22]: Should the DeletePolicy of this task be changed so it is not deleted?

File: src/shared/settingslist.h [hunk 16]
```diff
@@ -224,5 +224,27 @@ SETTING_STRING(token,        // getter
...
 )
 
+SETTING_STRING(sentryDSN,        // getter
+               setSentryDSN,     // setter
+               removeSentryDSN,  // remover
+               hasSentryDSN,     // has
+               "sentry/dsn",     // key
+               "",               // default value
+               false,            // user setting
+               false,            // remove when reset
```
Comment (vinocher) [inline src/shared/settingslist.h:234]: Is it better to remove when reset?

File: linux/debian/control [hunk 17]
```diff
@@ -27,7 +27,9 @@ Build-Depends: debhelper (>= 11.2),
                qmake6 (>=6.2.0~),
                qt6-base-dev (>=6.2.0~),
                qt6-base-dev-tools (>=6.2.0~),
+               qt6-base-private-dev (>=6.2.0~),
```
Comment (vinocher) [inline linux/debian/control:30]: For my info, could you please explain why these packages with private headers are needed?

**Comment by vinocher** [comment]
Just checking, are the changes in vpn\translations\i18n needed for this change?

File: src/shared/tasks/sentry/tasksentry.cpp [hunk 18]
```diff
@@ -76,12 +76,20 @@ void TaskSentry::run() {
...
+    return;
+  }
 
   request->requestInternal().setHeader(QNetworkRequest::ContentTypeHeader,
                                        "application/x-sentry-envelope");
-  request->requestInternal().setRawHeader("dsn",
-                                          Constants::SENTRY_DSN_ENDPOINT);
-  request->post(QUrl(Constants::SENTRY_ENVELOPE_INGESTION), m_envelope);
+  request->requestInternal().setRawHeader("dsn", dsn.toLocal8Bit());
+  request->post(QUrl(endpoint), m_envelope);
```
Comment (vinocher) [inline src/shared/tasks/sentry/tasksentry.cpp:92]: Should QUrl be validated (StrictMode parsing + IsValid) before the post?

File: src/shared/sentry/sentryadapter.cpp [hunk 19]
```diff
@@ -159,6 +171,7 @@ void SentryAdapter::declineCrashReporting() {
 }
 
 void SentryAdapter::captureQMLStacktrace(const char* description) {
+#ifndef UNIT_TEST
```
Comment (vinocher) [inline src/shared/sentry/sentryadapter.cpp:174]: Is this because we don't want to report crashes for unit tests?

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+    logger.error() << "DSN missing from Sentry info";
+    return false;
+  };
+
+  if (!obj.contains("endpoint")) {
+    logger.error() << "endpoint missing from Sentry info";
+    return false;
+  };
+  auto endpoint = obj["endpoint"].toString();
+  if (endpoint.isNull()) {
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:75]: Should QUrl be used here to parse the endpoint to check that the provided URL is valid?

File: .github/workflows/functional_tests.yaml (hunk 9 above)
Comment (strseb) [reply .github/workflows/functional_tests.yaml:27]: Good catch! added that.

File: linux/debian/control (hunk 17 above)
Comment (strseb) [reply linux/debian/control:30]: https://github.com/mozilla-mobile/mozilla-vpn-client/pull/7203

We currently have an influx of crashes that are triggered from "something" called in the js-engine. 
So in sentryadapter we added a crashandler that can hook into the JS stack and produce a mini stacktrace - that will be attached in a sentry report. 
Those api's are not public. Since this pr will enable us to also ship sentry on linux we know have a dependency on both private qt and qml headers :)

File: src/shared/sentry/sentryadapter.cpp (hunk 19 above)
Comment (strseb) [reply src/shared/sentry/sentryadapter.cpp:174]: No sadly this is me having a problem getting this to build on github CI. 
For an (for me) unknown reason, the compiler cannot find the private headers i.e  `include <QtQml/private/qqmlengine_p.h>` - which will cause this to fail if we build the tests. 
Given that we currently have no callers when running tests, i'd like to disable this and fix it once we actually need that.

File: src/shared/tasks/sentry/tasksentry.cpp (hunk 10 above)
Comment (strseb) [reply src/shared/tasks/sentry/tasksentry.cpp:78]: Not in this case - `NetworkRequest` is self-destructive after the request has finished :) 
https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/shared/networkrequest.h#L26

File: src/shared/tasks/sentryconfig/tasksentryconfig.h (hunk 11 above)
Comment (strseb) [reply src/shared/tasks/sentryconfig/tasksentryconfig.h:14]: I appreciate the callout :)

File: src/shared/tasks/sentryconfig/tasksentryconfig.h (hunk 15 above)
Comment (strseb) [reply src/shared/tasks/sentryconfig/tasksentryconfig.h:22]: Good catch! I've made it rescheudleable, given that this is not time critical :)

File: src/shared/settingslist.h (hunk 16 above)
Comment (strseb) [reply src/shared/settingslist.h:234]: Oh right, yes let's do that :)

**Comment by strseb** [comment]
> Just checking, are the changes in vpn\translations\i18n needed for this change?

Nope,  remove that :)

File: src/shared/tasks/sentry/tasksentry.cpp (hunk 10 above)
Comment (vinocher) [reply src/shared/tasks/sentry/tasksentry.cpp:78]: Thanks for pointing out the following in NetworkRequest will free the allocation:
connect(&m_timer, &QTimer::timeout, this, &NetworkRequest::maybeDeleteLater);
connect(m_reply, &QNetworkReply::finished, this,
          &NetworkRequest::maybeDeleteLater);

File: src/shared/tasks/sentryconfig/tasksentryconfig.h [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * @brief TaskSentryConfig fetches Sentry Configuration
+ *
+ * This Task will schedule a Network request to Guardians
+ * `crashreporting` endpoint to get info on where to
+ * post crashreports to.
+ *
+ * After a successful fetch, it will store the config
+ * in the Settings an Call SentryAdapter::init to
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.h:22]: Nit: Typo 'an'

File: src/shared/tasks/sentryconfig/tasksentryconfig.cpp [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+  if (!obj.contains("endpoint")) {
+    logger.error() << "endpoint missing from Sentry info";
+    return;
+  };
+  auto endpoint = obj["endpoint"].toString();
+  if (endpoint.isNull()) {
+    logger.error() << "Sentry provided endpoint is null";
+    return;
+  };
+  if (!QUrl(endpoint).isValid()) {
```
Comment (vinocher) [inline src/shared/tasks/sentryconfig/tasksentryconfig.cpp:75]: Should strict mode be used here? Also, is it sufficient to rely on the check here and remove the URL validity check in  TaskSentry::sendRequest?
